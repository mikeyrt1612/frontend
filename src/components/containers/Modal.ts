import { Dispatch } from 'redux'
import { connect } from 'react-redux'
import { createStructuredSelector } from 'reselect'

import { IRootState } from '@store/model'
import { isOpenSelector, modalConfigSelector } from '@store/modal/selectors'
import { closeModal } from '@store/modal/actions'
import Modal, { IStateProps, IDispatchProps } from '@components/organisms/Modal'

export const mapStateToProps = createStructuredSelector<IRootState, IStateProps>({
  isOpen: isOpenSelector,
  config: modalConfigSelector,
})

export const mapDispatchToProps = (dispatch: Dispatch): IDispatchProps => ({
  closeModal: () => { dispatch(closeModal()) },
})

export default connect(mapStateToProps, mapDispatchToProps)(Modal)
