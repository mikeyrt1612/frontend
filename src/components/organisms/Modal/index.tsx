import React, { Component, ReactNode } from 'react'
import { Button, Header, Modal as SemanticModal } from 'semantic-ui-react'
import isFunction from 'lodash/isFunction'

import { IModalConfig, IUpdateData } from '@store/modal/model'

export interface IStateProps {
  isOpen: boolean,
  config: IModalConfig,
}

export interface IDispatchProps {
  closeModal: () => void,
}

type IProps = IStateProps & IDispatchProps

interface IState {
  data: any,
  content: ReactNode,
}

class Modal extends Component<IProps, IState> {
  state = {
    data: undefined,
    content: undefined,
  }

  handleOpen = () => {
    const { config } = this.props

    let content: ReactNode
    if (isFunction(config.content)) {
      content = config.content(this.updateData)
    } else {
      content = config.content
    }

    this.setState({ content })
  }

  handleClose = () => {
    const { closeModal } = this.props
    closeModal()
    this.setState({ data: undefined })
  }

  updateData: IUpdateData<any> = (data) => {
    this.setState(
      ({ data: prevData }: IState) => ({
        data: {
          ...prevData,
          ...data,
        },
      }),
    )
  }

  isActionButtonDisabled = () => {
    const { config } = this.props
    const { data } = this.state

    if (isFunction(config.action.disabled)) {
      if (data === undefined) {
        return false
      }
      return config.action.disabled(data)
    }
    return config.action.disabled
  }

  render() {
    const { isOpen, config, closeModal } = this.props
    const { data, content } = this.state
    return (
      <SemanticModal
        dimmer="blurring"
        size="tiny"
        closeOnDimmerClick
        open={isOpen}
        onMount={this.handleOpen}
        onClose={this.handleClose}
        >
          <SemanticModal.Header>
            <Header icon={config.header.icon} content={config.header.label} />
          </SemanticModal.Header>
          <SemanticModal.Content content={content} />
          <SemanticModal.Actions>
            <Button
              content="Cancel"
              onClick={closeModal}
            />
            <Button
              primary
              disabled={this.isActionButtonDisabled()}
              content={config.action.label}
              onClick={() => {
                config.action.callback(data)
                closeModal()
              }}
            />
          </SemanticModal.Actions>
        </SemanticModal>
    )
  }
}

export default Modal
