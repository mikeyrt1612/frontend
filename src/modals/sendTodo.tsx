import React, { Fragment } from 'react'
import LabelPrimary from '@components/atoms/LabelPrimary'
import { IModalConfig } from '@store/modal/model'

export type ISendTodo = (recipientId: string) => void

export default (fullName: string, sendTodo: ISendTodo, id: string): IModalConfig => ({
  header: {
    label: 'Send Todo',
    icon: 'mail',
  },
  content: (
    <Fragment>
      <span>Are you sure you want to send </span>
      <LabelPrimary text={fullName} />
      <span> this todo?</span>
    </Fragment>
  ),
  action: {
    label: 'Send',
    callback: () => { sendTodo(id) },
  },
})
