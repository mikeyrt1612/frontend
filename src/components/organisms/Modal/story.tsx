import React, { Fragment, SyntheticEvent } from 'react'
import { storiesOf } from '@storybook/react'
import { Image, Checkbox, CheckboxProps } from 'semantic-ui-react'

import { IModalConfig } from '@store/modal/model'
import Modal from '.'

storiesOf('organisms/Modal', module)
  .add('string', () => (
    <Modal
      isOpen
      closeModal={() => {}}
      config={{
        header: {
          icon: 'trash alternate',
          label: 'Delete',
        },
        content: 'Are you sure you want to delete this?',
        action: {
          label: 'Delete',
          callback: () => {},
        },
      }}
    />
  ))
  .add('JSX', () => (
    <Modal
      isOpen
      closeModal={() => {}}
      config={{
        header: {
          icon: 'world',
          label: 'Travel to Mars',
        },
        content: (
          <Image src="https://amp.businessinsider.com/images/5b9235bb5c5e5254548b59f5-750-563.jpg"/>
        ),
        action: {
          label: 'Launch!',
          callback: () => {},
        },
      }}
    />
  ))
  .add('JSX with data', () => (
    <Modal
      isOpen
      closeModal={() => {}}
      config={{
        header: {
          icon: 'tasks',
          label: 'Select attributes',
        },
        content: (update) => {
          update({ creative: false, curious: false, cuddly: false })
          return (
            <Fragment>
              <Checkbox
                label="Creative"
                onChange={(event: SyntheticEvent, { checked }: CheckboxProps) => {
                  update({ creative: checked })
                }}
              />
              <br />
              <Checkbox
                label="Curious"
                onChange={(event: SyntheticEvent, { checked }: CheckboxProps) => {
                  update({ curious: checked })
                }}
              />
              <br />
              <Checkbox
                label="Cuddly"
                onChange={(event: SyntheticEvent, { checked }: CheckboxProps) => {
                  update({ cuddly: checked })
                }}
              />
            </Fragment>
          )
        },
        action: {
          label: 'Submit',
          callback: (data) => { console.log(data) },
        },
      } as IModalConfig<{
        creative: boolean,
        curious: boolean,
        cuddly: boolean,
      }>}
    />
  ))
  .add('disabled action', () => (
    <Modal
      isOpen
      closeModal={() => {}}
      config={{
        header: {
          icon: 'ban',
          label: 'Disabled action',
        },
        content: (update) => {
          update({ isDisabled: true })
          return (
            <Checkbox
              label="Disable?"
              defaultChecked
              onChange={(event: SyntheticEvent, { checked }: CheckboxProps) => {
                update({ isDisabled: checked })
              }}
            />
          )
        },
        action: {
          label: 'Submit',
          callback: () => {},
          disabled: ({ isDisabled }) => isDisabled,
        },
      } as IModalConfig<{ isDisabled: boolean }>}
    />
  ))
