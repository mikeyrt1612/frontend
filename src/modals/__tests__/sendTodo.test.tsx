import React, { Fragment, cloneElement } from 'react'
import { shallow, mount } from 'enzyme'
import renderer from 'react-test-renderer'
import { Button, Modal as SemanticModal } from 'semantic-ui-react'

import sendTodo from '../sendTodo'
import Modal from '@components/organisms/Modal'

const mockName = 'MOCK NAME'
const mockId = 'MOCK ID'

describe('sendTodo modal config', () => {
  test('should call the sendTodo callback with id when action button clicked', () => {
    const callback = jest.fn()
    const wrapper = shallow(
      <Modal
        isOpen
        config={sendTodo(mockName, callback, mockId)}
        closeModal={() => {}}
      />,
    )
    const actionButton = wrapper.findWhere(e =>
      e.type() === Button && e.prop('primary'),
    )
    actionButton.simulate('click')
    expect(callback).toBeCalledWith(mockId)
  })

  test('should have the content we expect', () => {
    const wrapper = mount(
      <Modal
        isOpen
        config={sendTodo(mockName, () => {}, mockId)}
        closeModal={() => {}}
      />,
    )
    const keyedContent = wrapper.find(SemanticModal.Content).children().children().map((e, i) =>
      cloneElement(e.getElement(), { key: i }),
    )
    const tree = renderer.create(
      <Fragment>
        {keyedContent}
      </Fragment>,
    ).toJSON()
    expect(tree).toMatchSnapshot()
    wrapper.unmount()
  })
})
