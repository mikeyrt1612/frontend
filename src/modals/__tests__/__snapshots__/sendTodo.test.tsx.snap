// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`sendTodo modal config should have the content we expect 1`] = `
Array [
  <span>
    Are you sure you want to send 
  </span>,
  <span
    className="sc-bdVaJa gzaUvJ"
  >
    MOCK NAME
  </span>,
  <span>
     this todo?
  </span>,
]
`;
